File: Atcoder/ABC348/D.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(int(1e5))

N, M = map(int, input().split())
board = []
for i in range(N):
    board.append(list(input().rstrip()))
    for j in range(M):
        if board[i][j] == "S":
            start = (i, j)
        elif board[i][j] == "T":
            goal = (i, j)
medicine = [[0] * M for _ in range(N)]
for medi in range(int(input())):
    r, c, e = map(int, input().split())
    medicine[r - 1][c - 1] = e

dx = [-1, 1,  0, 0]
dy = [0, 0, -1, 1]
dp = [[-1] * M for _ in range(N)]


def dfs(cx, cy, energy):
    if energy <= 0:
        return False
    for d in range(4):
        px = cx + dx[d]
        py = cy + dy[d]
        if px < 0 or px >= N or py < 0 or py >= M:
            continue
        if board[px][py] == "#":
            continue
        if (px, py) == goal:
            return True
        next_energy = max(energy - 1, medicine[px][py])
        if dp[px][py] >= next_energy:
            continue
        dp[px][py] = next_energy
        if dfs(px, py, next_energy):
            return True
    return False


dp[start[0]][start[1]] = medicine[start[0]][start[1]]
if dfs(start[0], start[1], medicine[start[0]][start[1]]):
    print("Yes")
else:
    print("No")

File: Atcoder/ABC348/D.cpp
#include <bits/stdc++.h>
using namespace std;

void fastio(){
    cin.tie(0);
    cout.tie(0);
    ios_base::sync_with_stdio(false);
}

int H, W, M;
vector<string> board;
vector<vector<int>> medicine;
vector<vector<int>> dp;
pair<int, int> start;
int dx[4] = {-1, 1, 0, 0}, dy[4] = {0, 0, -1, 1};

void solve(){
    cin >> H >> W;
    for(int i=0; i<H; i++){
        string s;
        cin >> s;
        board.push_back(s);
        for(int j=0; j<W; j++){
            if(s[j] == 'S') start = {i, j};
        }
    }
    medicine.resize(H, vector<int>(W, 0));
    dp.resize(H, vector<int>(W, -1));
    cin >> M;
    for(int i=0; i<M; i++){
        int r, c, e;
        cin >> r >> c >> e;
        medicine[r-1][c-1] = e;
    }
    queue<tuple<int, int, int>> PQ;
    PQ.push({medicine[start.first][start.second], start.first, start.second});
    dp[start.first][start.second] = medicine[start.first][start.second];
    while(!PQ.empty()){
        auto [energy, cx, cy] = PQ.front(); PQ.pop();
        if(dp[cx][cy] > energy) continue;
        if(energy == 0) continue;
        for(int d = 0; d<4; d++){
            int px = cx + dx[d], py = cy + dy[d];
            if(px < 0 || px >= H || py < 0 || py >= W) continue;
            if(board[px][py] == '#') continue;
            if(board[px][py] == 'T'){
                cout << "Yes" << '\n';
                return;
            }
            int next_energy = max(energy - 1, medicine[px][py]);
            if(dp[px][py] >= next_energy) continue;
            dp[px][py] = next_energy;
            PQ.push({next_energy, px, py});
        }
    }
    cout << "No" << '\n';
    return;
}

int main(){
    fastio();
    solve();
    return 0;
}